### PR TITLE
Export morphisms to JSON

### PR DESCRIPTION
**Pull request by TKuh**
This is very rudimentary right now.

A problem is, that the annotations in the JSON format are not optional:
https://github.com/Quantomatic/pyzx/blob/18e0a4d3004f5972afcb4a33de7d325df6c539f7/pyzx/graph/jsonparser.py#L62
(Even though it's stated in their documentation to be optional, it's probably too old...)

And this is a reason for failure:
```gap
gap tst/CategoryOfZXDiagrams.tst 

gap> ExportAsQGraph( H, "H" );
```

Produces a file `H.qgraph`:

```json
{
  "node_vertices": {
    "0": {
      "data": {
        "type": "input"
      }
    },
    "1": {
      "data": {
        "type": "H"
      }
    },
    "2": {
      "data": {
        "type": "output"
      }
    }
  },
  "undir_edges": {
    "E01": {
      "src": 0,
      "tgt": 1
    },
    "E12": {
      "src": 1,
      "tgt": 2
    }
  }
}
```
Trying to load it in Python:

```python
>>> import pyzx as zx
>>> with open("H.qgraph") as file:
...     json = file.read()
... 
>>> zx.Graph.from_json(json)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/tom/Software/Python/lib/python3.11/site-packages/pyzx/graph/base.py", line 474, in from_json
    return json_to_graph(js,cls.backend)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tom/Software/Python/lib/python3.11/site-packages/pyzx/graph/jsonparser.py", line 62, in json_to_graph
    c = attr['annotation']['coord']
        ~~~~^^^^^^^^^^^^^^
KeyError: 'annotation'

```
I have not tested setting the annotations to a fixed value (or something else), because I don't know if this is even desirable.
Also, I barely understand the various node/vertex types and which ones have to be chosen for the JSON. Right now, I'm only making a guess from examples I've found.

We have to talk about this in person next week, I would rather not have to dig through the ZX-Calculus, because that likely takes too much time.

### CODE REVIEW

**Comment by zickgraf** [comment]
Thanks for the progress so far, this already looks very promising :-)

**Comment by TKuh** [comment]
Some things I noticed:
1. `Hadamard` nodes can have an attribute `is_edge` (see [here](https://github.com/Quantomatic/pyzx/blob/a215056115e7103c5bc492cef1d21bc121ff1dad/pyzx/graph/jsonparser.py#L59))
2. There are also `W_input`/`W_output` node vertices. (see [here](https://github.com/Quantomatic/pyzx/blob/a215056115e7103c5bc492cef1d21bc121ff1dad/pyzx/graph/jsonparser.py#L74-L75))

I don't know if this is important, I just wanted to point this out to you.

Other than that, the JSON's of the morphisms from `CategoryOfZXDiagrams.tst` are all accepted by PyZX. :)

```gap
gap> ExportAsQGraph( H, "H" );
gap> ExportAsQGraph( X_1_1, "X" );
gap> ExportAsQGraph( Z_1_1, "Z" );

gap> precomp := PreCompose( ev, coev );
gap> pre2 := PreCompose( coev, ev );
gap> ExportAsQGraph( precomp, "precomp" );
gap> ExportAsQGraph( pre2, "precomp2" );
```
```python
import pyzx as zx

for name in [ "X.qgraph", "Z.qgraph", "H.qgraph", "precomp.qgraph", "precomp2.qgraph" ]:
     with open( name ) as file:
          json = file.read()
          zx.Graph.from_json(json)
```

**Review by zickgraf — CHANGES_REQUESTED** [review]
Thanks, that already looks very nice :-) Some code comments below. A functional comment will follow below.

File: gap/Tools.gi [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+                   undir_edges := undir_edges );
+    
+    qgraph := GapToJsonString( qgraph );
+    
+    file := OutputTextFile( Concatenation( filename, ".qgraph" ), false );;
+    
+    SetPrintFormattingStatus( file, false );
+    
+    PrintTo( file, qgraph );
+
```
Comment (zickgraf) [inline gap/Tools.gi:195]: Please use `CloseStream` to close the output stream here (see documentation of `OutputTextFile`).

Comment (zickgraf) [reply gap/Tools.gi:195]: Or use `FileString` which should be more convenient anyway :-) (Yes, GAP and homalg have lots of different IO strategies :D )

File: gap/Tools.gi [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+    local pair, labels, edges, pos, edge_positions, new_edge, edge_1, edge_2,
+          remaining_indices, wire_vertices, node_vertices, undir_edges, edge, source, target,
+          edge_name, qgraph, file;
+    
+    pair := MorphismDatum( phi );
+    
+    labels := ShallowCopy( pair[1] );
+    edges := ShallowCopy( pair[2] );
+    
+    # remove neutral nodes
```
[the remark this answers is not in the document]
Comment (TKuh) [reply gap/Tools.gi]: Done.

File: gap/Tools.gi [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+                                             input := false,
+                                             output := true );
+        
+        elif labels[ pos ] = "input_output" then
+        
+           wire_vertices.( pos - 1 ) := rec( annotation := rec( coord := [ 0, 0 ] ),
+                                             input := true,
+                                             output := true );
+        
+        # TODO: neutral nodes?
```
[the remark this answers is not in the document]
Comment (TKuh) [reply gap/Tools.gi]: Nothing, I forgot to remove it.

File: gap/Tools.gi [hunk 4]
```diff
@@ -0,0 +1,196 @@
+# SPDX-License-Identifier: GPL-2.0-or-later
+# ZXCalculusForCAP: The category of ZX-diagrams
+#
+# Implementations
+#
+
+InstallGlobalFunction( ExportAsQGraph,
```
[the remark this answers is not in the document]
Comment (TKuh) [reply gap/Tools.gi]: Thanks, done.

File: gap/Tools.gi (hunk 1 above)
Comment (TKuh) [reply gap/Tools.gi]: Thank you, yes, that is confusing (and not so easy to find). :D
Also done.

File: gap/Tools.gi [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+    od;
+    
+    undir_edges := rec( );
+    
+    for edge in edges do
+    
+        source := edge[1];
+        target := edge[2];
+        
+        edge_name := Concatenation( "E", String( source ), String( target ) );
```
[the remark this answers is not in the document]
Comment (TKuh) [reply gap/Tools.gi]: Thanks a lot for that catch! It's fixed now.

**Comment by TKuh** [comment]
> All of this already seems to work quite well, but PyZX has a weird autodetection of inputs and outputs from the coordinates (input nodes are drawn on the left, output nodes on the right). This autodetection seems to kick in when using the json import, and then the information about inputs and outputs is lost. 

It turned out that this information is not read, because `json_to_graph` expects the `input/output` field to be inside the `annotation` field.  Now it's imported correctly. If you prefer, I can also adjust the coordinates.

Another thing is this:
```gap
gap ./tst/CategoryOfZXDiagrams.tst

gap> precomp := PreCompose( ev, coev );
gap> ExportAsQGraph( precomp, "precomp" );
```
The above gives the following edges:
```json
"undir_edges": {
    "e0": {
      "src": "0",
      "tgt": "5"
    },
    "e1": {
      "src": "1",
      "tgt": "4"
    },
    "e2": {
      "src": "2",
      "tgt": "3"
    },
    "e3": {
      "src": "6",
      "tgt": "11"
    },
    "e4": {
      "src": "7",
      "tgt": "10"
    },
    "e5": {
      "src": "8",
      "tgt": "9"
    }
  },
```

After `print(zx.Graph.from_json(json).to_json())` some edges have their sources and targets swapped:

```json
"undir_edges": {
    "e0": {
      "src": "0",
      "tgt": "5"
    },
    "e1": {
      "src": "1",
      "tgt": "4"
    },
    "e2": {
      "src": "10",
      "tgt": "7"
    },
    "e3": {
      "src": "11",
      "tgt": "6"
    },
    "e4": {
      "src": "2",
      "tgt": "3"
    },
    "e5": {
      "src": "8",
      "tgt": "9"
    }
  },
```
I suspect it's because of this call
https://github.com/Quantomatic/pyzx/blob/a215056115e7103c5bc492cef1d21bc121ff1dad/pyzx/graph/jsonparser.py#L152

From the documentation of `‎add_edge_table‎`: "[...] The method selectively adds or removes edges to produce that ZX diagram which would result from adding (#edges, #h-edges), and then removing all parallel edges using Hopf/spider laws."

File: gap/Tools.gi [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+                                                                     input := false,
+                                                                     output := true ) );
+            
+            elif labels[ pos ] = "input_output" then
+            
+                wire_vertices.( pos - 1 ) := rec( annotation := rec( coord := [ 0, 0 ],
+                                                                     input := true,
+                                                                     output := true ) );
+            
+            fi;
```
Comment (zickgraf) [inline gap/Tools.gi:62]: Please employ defensive programming, i.e. explicitly throw an error in the else case. We do not want to silently miss edges if something is wrong.

File: gap/Tools.gd [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+#
+#! @Chapter Tools
+
+#! @Section Exporting ZX-diagrams
+
+#! @Description
+#!   Takes a morphism in a category of ZX-diagrams and a filename **file**
+#!   and exports the morphism as **file**.qgraph.
+#!   See https://github.com/Quantomatic/quantomatic/blob/stable/docs/json_formats.txt
+#!   for an overview of the qgraph format.
```
Comment (zickgraf) [inline gap/Tools.gd:14]: Please mention that this function is only available if the package `json` is.

File: gap/Tools.gi [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+            
+        od;
+        
+        undir_edges := rec( );
+        
+        for edge in edges do
+        
+            edge_name := Concatenation( "e", String( Position( edges, edge ) - 1 ) );
```
Comment (zickgraf) [inline gap/Tools.gi:70]: This still does not work for multiple edges between the same nodes, or does it?

**Review by zickgraf — CHANGES_REQUESTED** [review]
> It turned out that this information is not read, because json_to_graph expects the input/output field to be inside the annotation field. Now it's imported correctly. If you prefer, I can also adjust the coordinates.

Perfect, then we do not need the coordinates.

> After print(zx.Graph.from_json(json).to_json()) some edges have their sources and targets swapped:

The semantics do not care about the direction of edges, so that should be fine.

**Comment by zickgraf** [comment]
Ah, the CI does not run because you are a first-time contributor. Usually there is a button to trigger the CI manually, but I cannot find it. Please make sure `make ci-test` succeeds.

File: gap/Tools.gd (hunk 7 above)
Comment (TKuh) [reply gap/Tools.gd:15]: Done.

File: gap/Tools.gi (hunk 8 above)
Comment (TKuh) [reply gap/Tools.gi]: Thank you, I overestimated `Position`. Now it works and I also added a morphism with two edges in the test.

**Comment by TKuh** [comment]
> Ah, the CI does not run because you are a first-time contributor. Usually there is a button to trigger the CI manually, but I cannot find it. Please make sure `make ci-test` succeeds.

I get this Julia error at the end:
```julia
git clone https://github.com/zickgraf/ZXCalculusForCAP.jl.git ~/.julia/dev/ZXCalculusForCAP
Cloning into '/home/tom/.julia/dev/ZXCalculusForCAP'...
remote: Enumerating objects: 24, done.
remote: Counting objects: 100% (24/24), done.
remote: Compressing objects: 100% (21/21), done.
remote: Total 24 (delta 1), reused 24 (delta 1), pack-reused 0
Receiving objects: 100% (24/24), 21.28 KiB | 807.00 KiB/s, done.
Resolving deltas: 100% (1/1), done.
~/.gap/pkg/PackageJanitor/gap_to_julia CAP
rm: cannot remove '/home/tom/.julia/dev/CAP/docs/src/cap20.tst.autogen.md': No such file or directory
make: *** [makefile:71: test-gap_to_julia] Error 1
```

**Comment by zickgraf** [comment]
I think I will take over at this point: I will push some cleanups and then work on some subtleties for PyZX. PyZX has more quirks then expected, and fixing them on the fly is more efficient than communicating back and forth.

Thanks a lot for the work!

**Review by zickgraf — APPROVED** [review]

**Pull request merged**